File: LeetCode/Medium/0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func flatten(root *TreeNode)  {
    dummy := &TreeNode{}
    
    postOrder := func(node *TreeNode){}
    postOrder = func(node *TreeNode){
        if node == nil { return }
        postOrder(node.Right)
        postOrder(node.Left)
        node.Right = dummy.Right
        node.Left = nil
        dummy.Right = node
    }
    
    postOrder(root)
}

File: LeetCode/Medium/0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode dummy;
    public void flatten(TreeNode root) {
        dummy = new TreeNode();
        
        postOrder(root);
    }
    
    private void postOrder(TreeNode node){
        if (node == null) return;
        
        postOrder(node.right);
        postOrder(node.left);
        node.right = dummy.right;
        node.left = null;
        dummy.right = node;
    }
}